subsys/net/lib/wifi_credentials/wifi_credentials_backend_settings.c

```diff
@@ -0,0 +1,143 @@
+ */
+static int settings_set(const char *key, size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	int ret = 0;
+
+	/* key validation */
+	if (!key) {
+		LOG_ERR("Settings error: no key");
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&wifi_credentials_mutex, K_FOREVER);
+
+	int idx = atoi(key);
+
+	if ((idx == 0 && strcmp(key, "0") != 0) ||
+	     idx >= CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		LOG_ERR("Settings error: index too large");
+		ret = -EINVAL;
+		goto exit;
+	}
+
+	if (len < sizeof(struct wifi_credentials_header)) {
+		LOG_ERR("Settings error: invalid settings length");
+		ret = -EINVAL;
+		goto exit;
+	}
+
+	uint8_t buf[ENTRY_MAX_LEN];
+	size_t length_read = read_cb(cb_arg, buf, ARRAY_SIZE(buf));
+
+	/* value validation */
+	if (length_read < len) {
+		LOG_ERR("Settings error: entry incomplete");
+		ret = -ENODATA;
+		goto exit;
+	}
+
+	wifi_credentials_cache_ssid(idx, (struct wifi_credentials_header *) buf);
+
+exit:
+	k_mutex_unlock(&wifi_credentials_mutex);
+	return ret;
+}
+
+int wifi_credentials_store_entry(size_t idx, const void *buf, size_t buf_len)
+{
+	char settings_name_buf[16] = {0};
+
+	int ret = snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+
+	if (ret < 0 || ret == ARRAY_SIZE(settings_name_buf)) {
+		LOG_ERR("WiFi credentials settings key could not be generated, idx: %d", idx);
+		return -EFAULT;
+	}
+
+	return settings_save_one(settings_name_buf, buf, buf_len);
+}
+
+int wifi_credentials_delete_entry(size_t idx)
+{
+	char settings_name_buf[16] = {0};
+
+	int ret = snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+
+	if (ret < 0 || ret == ARRAY_SIZE(settings_name_buf)) {
+		LOG_ERR("WiFi credentials settings key could not be generated, idx: %d", idx);
+		return -EFAULT;
+	}
+
+	return settings_delete(settings_name_buf);
+}
+
+/* Type of the callback argument used in the function below. */
+struct zephyr_settings_backend_load_cb_arg {
+	uint8_t *buf;
+	size_t buf_len;
+};
+
+/* This callback function is used to retrieve credentials on demand. */
+static int zephyr_settings_backend_load_cb(const char *key, size_t len,
+			   settings_read_cb read_cb,
+			   void *cb_arg, void *param)
+{
+	struct zephyr_settings_backend_load_cb_arg *arg = param;
+
+	if (len < sizeof(struct wifi_credentials_header)) {
+		LOG_ERR("Settings error: invalid settings length");
+		return -EINVAL;
+	}
+
+	size_t length_read = read_cb(cb_arg, arg->buf, arg->buf_len);
+
+	/* value validation */
+	if (length_read < len) {
+		LOG_ERR("Settings error: entry incomplete");
+		return -ENODATA;
+	}
+
+	return 0;
+}
+
+int wifi_credentials_load_entry(size_t idx, void *buf, size_t buf_len)
+{
+	struct zephyr_settings_backend_load_cb_arg arg = {
+		.buf = buf,
+		.buf_len = buf_len
+	};
+
+	/* Browse through the settings entries with custom callback to load the whole entry. */
+	return settings_load_subtree_direct("wifi_cred", zephyr_settings_backend_load_cb, &arg);
+}
```

Comment:
`idx` is not used, so this will probably load the last entry

Reply:
fixed